Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py

```diff
@@ -1,6 +1,1031 @@
+            'text_length': text_length,
+            'num_of_words_subject': num_of_words_subject,
+            'subject_length': subject_length,
+            'ending_dots': ending_dots,
+            'ending_explanation': ending_explanation,
+            'ending_question': ending_question,
+            'number_of_lines': number_of_lines
+            }
+
+
+def get_characters_features(text):
+    characters_count = {}
+    for c in string.printable:
+        if c in ['[', ']', '<']:
+            continue
+        count = text.count(c)
+        if count > 0:
+            characters_count[c] = count
+    return characters_count
+
+
+def get_url_features(email_body, email_html, soup):
+    http_regex = r'http://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
+    https_regex = r'https://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
+    http_urls = re.findall(http_regex, email_body)
+    https_urls = re.findall(https_regex, email_body)
+    embedded_urls = [a['href'] for a in soup.findAll('a')]
+    all_urls = http_urls + https_urls + embedded_urls
+    all_urls_lengths = [len(u) for u in all_urls]
+    average_url_length = 0 if len(all_urls) == 0 else sum(all_urls_lengths) / len(all_urls)
+    min_url_length = 0 if len(all_urls_lengths) == 0 else min(all_urls_lengths)
+    max_url_length = 0 if len(all_urls_lengths) == 0 else max(all_urls_lengths)
+    shortened_urls_count = len([u for u in all_urls if any(shortened_u in u for shortened_u in SHORTENED_DOMAINS)])
+    drive_count = len([u for u in all_urls if any(drive in u for drive in DRIVE_URL_KEYWORDS)])
+    return {
+        'http_urls_count': len(http_urls),
+        'https_urls_count': len(https_urls),
+        'embedded_urls_count': len(embedded_urls),
+        'all_urls_count': len(all_urls),
+        'average_url_length': average_url_length,
+        'min_url_length': min_url_length,
+        'max_url_length': max_url_length,
+        'shortened_urls_count': shortened_urls_count,
+        'drive_count': drive_count
+    }
+
+
+def get_html_features(soup):
+    global HTML_TAGS
+    html_counter = Counter([tag.name for tag in soup.find_all()])
+    for t in HTML_TAGS:
+        html_counter[t] = 0 if t not in html_counter else html_counter[t]
+    html_counter = {k: v for k, v in html_counter.items() if k in HTML_TAGS}
+    return html_counter
+
+
+def get_ml_features(tokenized_text):
+    global EMBEDDING_DICT
+    if EMBEDDING_DICT is None:
+        with open('/var/glove_top_20k.p', 'rb') as file:
+            EMBEDDING_DICT = pickle.load(file)
+    vectors = [EMBEDDING_DICT[w] for w in tokenized_text if w in EMBEDDING_DICT]
+    if len(vectors) == 0:
+        mean_vector = np.zeros(50)
+    else:
+        mean_vector = np.mean(vectors, axis=0)
+    res = {'glove_avg_{}'.format(i): mean_vector[i].item() for i in range(len(mean_vector))}
+    return res
+
+
+def get_header_value(email_headers, header_name, index=0, ignore_case=False):
```

Comment:
I think it terms of efficiency - it's worth to build email headers dict (key->value) once in `get_headers_features` method and pass that every time to this function 
here you iterate the headers list everytime you get header

Reply:
reason for that is mainly the for the Received headers which appear multiple times